src/plan/markcompact/global.rs

```diff
@@ -0,0 +1,210 @@
+// Add
+use crate::plan::markcompact::mutator::ALLOCATOR_MAPPING;
+use crate::plan::AllocationSemantics;
+use crate::plan::Plan;
+use crate::plan::PlanConstraints;
+use crate::policy::markcompactspace::MarkCompactSpace;
+use crate::policy::space::Space;
+use crate::scheduler::gc_work::*; // Add
+use crate::scheduler::*; // Modify
+use crate::util::alloc::allocators::AllocatorSelector;
+#[cfg(not(feature = "global_alloc_bit"))]
+use crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC;
+use crate::util::heap::layout::heap_layout::Mmapper;
+use crate::util::heap::layout::heap_layout::VMMap;
+use crate::util::heap::layout::vm_layout_constants::{HEAP_END, HEAP_START};
+use crate::util::heap::HeapMeta;
+use crate::util::heap::VMRequest;
+use crate::util::metadata::side_metadata::{SideMetadataContext, SideMetadataSanity};
+use crate::util::opaque_pointer::*;
+use crate::util::options::UnsafeOptionsWrapper;
+use crate::vm::ActivePlan;
+use crate::vm::ObjectModel;
+use crate::vm::VMBinding;
+use enum_map::EnumMap;
+use std::sync::Arc;
+
+// pub const ALLOC_MARKCOMPACT: AllocationSemantics = AllocationSemantics::Default; // Add
+
+pub struct MarkCompact<VM: VMBinding> {
+    pub mc_space: MarkCompactSpace<VM>,
+    pub common: CommonPlan<VM>,
+}
+
+pub const MARKCOMPACT_CONSTRAINTS: PlanConstraints = PlanConstraints {
+    moves_objects: true,
+    gc_header_bits: 2,
+    gc_header_words: 1,
+    gc_extra_header_words: 1,
+    num_specialized_scans: 2,
+    ..PlanConstraints::default()
+};
+
+impl<VM: VMBinding> Plan for MarkCompact<VM> {
+    type VM = VM;
+
+    fn constraints(&self) -> &'static PlanConstraints {
+        &MARKCOMPACT_CONSTRAINTS
+    }
+
+    fn create_worker_local(
+        &self,
+        tls: VMWorkerThread,
+        mmtk: &'static MMTK<Self::VM>,
+    ) -> GCWorkerLocalPtr {
+        let mut c = NoCopy::new(mmtk);
+        c.init(tls);
+        GCWorkerLocalPtr::new(c)
+    }
+
+    fn gc_init(
+        &mut self,
+        heap_size: usize,
+        vm_map: &'static VMMap,
+        scheduler: &Arc<GCWorkScheduler<VM>>,
+    ) {
+        self.common.gc_init(heap_size, vm_map, scheduler);
+        self.mc_space.init(vm_map);
+    }
+
+    fn base(&self) -> &BasePlan<VM> {
+        &self.common.base
+    }
+
+    fn common(&self) -> &CommonPlan<VM> {
+        &self.common
+    }
+
+    fn prepare(&mut self, _tls: VMWorkerThread) {
+        self.common.prepare(_tls, true);
+        self.mc_space.prepare();
+    }
+
+    fn release(&mut self, _tls: VMWorkerThread) {
+        self.common.release(_tls, true);
+        self.mc_space.release();
+    }
+
+    fn get_allocator_mapping(&self) -> &'static EnumMap<AllocationSemantics, AllocatorSelector> {
+        &*ALLOCATOR_MAPPING
+    }
+
+    fn schedule_collection(&'static self, scheduler: &GCWorkScheduler<VM>) {
+        self.base().set_collection_kind::<Self>(self);
+        self.base().set_gc_status(GcStatus::GcPrepare);
+        scheduler.work_buckets[WorkBucketStage::Unconstrained]
```

Comment:
```CommonPlan.schedule_common() ``` will cause panic. It seems that finalization will perform a closure on dead objects. This will trigger assertation failure in markcompact.

Reply:
What is the assertion failure?

Reply:
```
debug_assert!(
    crate::util::alloc_bit::is_alloced(object),
    "{:x}: alloc bit not set",
    object
);
```
in ```trace_mark_object```. 
```is_live``` is implemented as 
```
fn is_live(&self, object: ObjectReference) -> bool {
    alloc_bit::is_alloced(object)
}
```

Reply:
Also finalization does not work with the current implementation of ```MarkingProcessEdges```

Reply:
Are you using your forward trace type for finalization? `schedule_common()` assumes there is one `ProcessEdgesType` per each GC. Clearly that is not the case for mark compact. If you can use `schedule_common()`, that would be better. But if you can't, please make sure you actually schedule those common work in `schedule_common()`. We can do the refactoring later.

Reply:
Then the question comes back to finalization. Does it perform a closure on dead objects? If so, I guess I need to either update the current ProcessEdges implementation or add a new one

Reply:
It does. It will call `trace_object()` to get the new object reference so we can update the object reference in the finalizable processor in case the object has been moved. I checked the Java MMTk, it uses the forward trace for finalization - you can check the implementation there.